ros/lib/utils.py

```diff
@@ -185,3 +193,61 @@ def log_request(self, code='-', size='-'):
             return
         else:
             super().log_request(code, size)
+
+
+def generate_highlight_description(instance_type):
+    if instance_type not in INSTANCE_TYPES.keys():
+        # logging lookup failure on Prometheus
+        ec2_instance_lookup_failures.labels(reporter='API Events').inc()
+        return 'NA'
+
+    cpu_type = INSTANCE_TYPES[instance_type]['extra']['physicalProcessor']
+    num_vcpus = INSTANCE_TYPES[instance_type]['extra']['vcpu']
+    ram_gb = INSTANCE_TYPES[instance_type]['extra']['memory']
+    cloud_regions = INSTANCE_TYPES[instance_type]['extra']['regionCode']
+    cloud = 'AWS'  # Update when multi cloud provider support is added
+
+    description_text = f'{cpu_type} instance with {num_vcpus} vCPUs ' \
+                       f'and {ram_gb} of RAM, running on {cloud} ' \
+                       f'{cloud_regions} regions'
+
+    return description_text
```

Comment:
```suggestion
   system_instance_type = INSTANCE_TYPES.get(instance_type)
    if system_instance_type is None:
        # logging lookup failure on Prometheus
        ec2_instance_lookup_failures.labels(reporter='API Events').inc()
        return 'NA'
    instance_type_extra = system_instance_type.get('extra')
    
    return ("{cpu_type} instance with {num_vcpus} vCPUs " \
                "and {ram_gb} of RAM, running on {cloud}" \
                "{cloud_regions} regions").format(cpu_type=instance_type_extra.get('physicalProcessor'),
                num_vcpus=instance_type_extra.get('vcpu'), ram_gb=instance_type_extra.get('memory'),
                cloud_regions=instance_type_extra.get('regionCode'), cloud= 'AWS' ) # Update when multi cloud provider support is added

```

Instead of reading object multiple times from `INSTANCE_TYPES`, will it be more efficient to capture respective instance_type object in single variable?

Reply:
@kgaikwad I ran some profiling with the suggested change,

Suggested change:

```
Timer unit: 1e-06 s
Total time: 8e-06 s
File: /home/nik/workspace/forked/ros-backend/ros/lib/utils.py
Function: generate_highlight_description at line 202
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   202                                           def generate_highlight_description(instance_type):
   203         1          1.0      1.0     12.5      system_instance_type = INSTANCE_TYPES.get(instance_type)
   204         1          1.0      1.0     12.5      if system_instance_type is None:
   205                                                   # logging lookup failure on Prometheus
   206                                                   ec2_instance_lookup_failures.labels(reporter='API Events').inc()
   207                                                   return 'NA'
   208         1          1.0      1.0     12.5      instance_type_extra = system_instance_type.get('extra')
   209                                           
   210         2          3.0      1.5     37.5      description_text = "{cpu_type} instance with {num_vcpus} vCPUs and {ram_gb} of RAM, running on {cloud}{cloud_regions} regions".format(
   211         1          0.0      0.0      0.0          cpu_type=instance_type_extra.get('physicalProcessor'),
   212         1          1.0      1.0     12.5          num_vcpus=instance_type_extra.get('vcpu'),
   213         1          0.0      0.0      0.0          ram_gb=instance_type_extra.get('memory'),
   214         1          0.0      0.0      0.0          cloud_regions=instance_type_extra.get('regionCode'),
   215         1          1.0      1.0     12.5          cloud='AWS')  # Update when multi cloud provider support is added
   216                                           
   217         1          0.0      0.0      0.0      return description_text
```

Existing change:

```
Timer unit: 1e-06 s
Total time: 8e-06 s
File: /home/nik/workspace/forked/ros-backend/ros/lib/utils.py
Function: generate_highlight_descriptione at line 220
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   220                                           def generate_highlight_descriptione(instance_type):
   221         1          5.0      5.0     62.5      if instance_type not in INSTANCE_TYPES.keys():
   222                                                   # logging lookup failure on Prometheus
   223                                                   ec2_instance_lookup_failures.labels(reporter='API Events').inc()
   224                                                   return 'NA'
   225                                           
   226         1          1.0      1.0     12.5      cpu_type = INSTANCE_TYPES[instance_type]['extra']['physicalProcessor']
   227         1          0.0      0.0      0.0      num_vcpus = INSTANCE_TYPES[instance_type]['extra']['vcpu']
   228         1          0.0      0.0      0.0      ram_gb = INSTANCE_TYPES[instance_type]['extra']['memory']
   229         1          0.0      0.0      0.0      cloud_regions = INSTANCE_TYPES[instance_type]['extra']['regionCode']
   230         1          0.0      0.0      0.0      cloud = 'AWS'  # Update when multi cloud provider support is added
   231                                           
   232         1          1.0      1.0     12.5      description_text = f'{cpu_type} instance with {num_vcpus} vCPUs ' \
   233                                                                  f'and {ram_gb} of RAM, running on {cloud} ' \
   234                                                                  f'{cloud_regions} regions'
   235                                           
   236         1          1.0      1.0     12.5      return description_text
```
As we can  see the total time remains roundabout the same, here are the key points

1.  `system_instance_type = INSTANCE_TYPES.get(instance_type)` is a good suggestion, the `in` lookup is expensive
2.  Fetching the `instance type['extra']` multiple times is not a problem as Python loads it only once in memory
3.  IMHO we should push towards f-strings going forward instead of format, as it is faster and easier to read

Here is the  profiling including the suggestion,
```
Timer unit: 1e-06 s
Total time: 6e-06 s
File: /home/nik/workspace/forked/ros-backend/ros/lib/utils.py
Function: generate_highlight_descriptione at line 220
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   220                                           def generate_highlight_descriptione(instance_type):
   221         1          2.0      2.0     33.3      _instance_type = INSTANCE_TYPES.get(instance_type, None)
   222         1          1.0      1.0     16.7      if _instance_type is None:
   223                                                   # logging lookup failure on Prometheus
   224                                                   ec2_instance_lookup_failures.labels(reporter='API Events').inc()
   225                                                   return 'NA'
   226                                           
   227         1          1.0      1.0     16.7      cpu_type = _instance_type['extra']['physicalProcessor']
   228         1          0.0      0.0      0.0      num_vcpus = _instance_type['extra']['vcpu']
   229         1          0.0      0.0      0.0      ram_gb = _instance_type['extra']['memory']
   230         1          0.0      0.0      0.0      cloud_regions = _instance_type['extra']['regionCode']
   231         1          0.0      0.0      0.0      cloud = 'AWS'  # Update when multi cloud provider support is added
   232                                           
   233         1          1.0      1.0     16.7      description_text = f'{cpu_type} instance with {num_vcpus} vCPUs ' \
   234                                                                  f'and {ram_gb} of RAM, running on {cloud} ' \
   235                                                                  f'{cloud_regions} regions'
   236                                           
   237         1          1.0      1.0     16.7      return description_text
```
I'll push the code shortly, meanwhile please feel free to share your thoughts

Reply:
Change pushed